### PR TITLE
Allow set custom podium for EnderDragon

### LINKED ISSUE #7693: Add option to override EnderDragon origin / target location
### Is your feature request related to a problem?

I'm currently working on a project, which requires multiple simultaneous EnderDragon fights (not a real DragonBattle, just killing a "manually" spawned entity). Since the EnderDragon pathfinding is hardcoded to be centered around [0,0] the easiest solution for me was just loading multiple worlds. But now I'm trying to expand on the whole concept, and it would be easier to just have multiple arenas in the same world instead of loading a new world for each fight. But that would require having multiple EnderDragons with different origin locations.

### Describe the solution you'd like.

Some possibility to define an override for the origin / target location of an EnderDragon entity.

Maybe something like:

- EnderDragon#setOrigin(@ Nullable Location origin)

which would override the default origin. Setting this value to null could restore the default behavior. (Which is targeting the EndPodiumFeature if I understood the dragon phase pathfinding correctly)

### Describe alternatives you've considered.

Modifying the pathfinding on my own and recreating default behavior, since I want 1:1 default AI, just not centered at [0,0].

### Other

_No response_

### PR DESCRIPTION
**Pull request by Doc94**
This PR resolve #7693 adding a API for set a custom podium this is handle for the most of logic for the EnderDragon Fight.

### CODE REVIEW

File: patches/server/0889-Allow-change-origin-for-EnderDragon.patch [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderDragon.java
+@@ -79,4 +79,16 @@ public class CraftEnderDragon extends CraftMob implements EnderDragon {
+     public int getDeathAnimationTicks() {
+         return this.getHandle().dragonDeathTime;
+     }
++
++    // Paper start
++    @Override
++    public org.bukkit.Location getOrigin() {
++        return null;
```
Comment (Owen1212055) [inline patches/server/0889-Allow-change-origin-for-EnderDragon.patch]: You need to call the NMS methods in here.

Comment (Doc94) [reply patches/server/0889-Allow-change-origin-for-EnderDragon.patch]: i keep this for the final and forget add when commit, xd

Added logic

**Comment by Doc94** [comment]
Ok i make a few changes because i forget the logic about `Heightmap.Types` then now this PR allow change the position of "Podium" and in base of this the rest of NMS use this for apply the logic for the differents `Heightmap.Types` for keep the logic from vanilla about the relative position.

File: patches/server/0889-Allow-change-podium-for-EnderDragon.patch [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
++    // Paper start
++    @Override
++    public org.bukkit.Location getPodium() {
++        net.minecraft.core.BlockPos blockPosOrigin = this.getHandle().getPodium();
++        return new org.bukkit.Location(getWorld(), blockPosOrigin.getX(), blockPosOrigin.getY(), blockPosOrigin.getZ());
++    }
++
++    @Override
++    public void setPodium(org.bukkit.Location location) {
++        if (location != null) {
```
Comment (kennytv) [inline patches/server/0889-Allow-change-podium-for-EnderDragon.patch]: Wrong way around?

File: patches/server/0889-Allow-change-podium-for-EnderDragon.patch [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
++            this.getHandle().setPodium(null);
++        } else {
++            if (location.getWorld() != null) {
```
Comment (kennytv) [inline patches/server/0889-Allow-change-podium-for-EnderDragon.patch]: Validate.isTrue(location.getWorld() == nul || ...

File: patches/api/0377-Allow-change-podium-for-EnderDragon.patch [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
++    /**
++     * Get the location considered by the dragon how podium.
++     *
++     * @return the podium location of the dragon
++     */
++    @NotNull
++    org.bukkit.Location getPodium();
++
++    /**
++     * Sets the location of podium for the dragon.
```
Comment (kennytv) [inline patches/api/0377-Allow-change-podium-for-EnderDragon.patch]:
```suggestion
+     * Sets the location of the podium for the ender dragon.
```

File: patches/api/0377-Allow-change-podium-for-EnderDragon.patch [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
++++ b/src/main/java/org/bukkit/entity/EnderDragon.java
+@@ -104,4 +104,22 @@ public interface EnderDragon extends ComplexLivingEntity, Boss, Mob {
+      * @return this dragon's death animation ticks
+      */
+     int getDeathAnimationTicks();
++
++    // Paper start
++
++    /**
++     * Get the location considered by the dragon how podium.
```
Comment (kennytv) [inline patches/api/0377-Allow-change-podium-for-EnderDragon.patch]:
```suggestion
+     * Get the podium location used by the ender dragon.
```

File: patches/api/0377-Allow-change-podium-for-EnderDragon.patch [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
++     *
++     * @return the podium location of the dragon
++     */
++    @NotNull
++    org.bukkit.Location getPodium();
++
++    /**
++     * Sets the location of podium for the dragon.
++     *
++     * @param location the location of the podium or null for use the default podium (Exit Portal of the END)
```
Comment (kennytv) [inline patches/api/0377-Allow-change-podium-for-EnderDragon.patch]:
```suggestion
+     * @param location the location of the podium or null to use the default podium location (exit portal of the end)
```

**Review by kennytv — CHANGES_REQUESTED** [review]

File: patches/server/0889-Allow-change-podium-for-EnderDragon.patch (hunk 2 above)
Comment (Doc94) [reply patches/server/0889-Allow-change-podium-for-EnderDragon.patch]: im pretty sure i use ==.. maybe a confusion when WSL still not opening popups :/ fixed

**Review by kennytv — APPROVED** [review]
For future reference; don't merge master, either rebase, or just don't until asked to rebase, else we can do it on merging

**Pull request merged**